pennylane/tape/qscript.py

```diff
@@ -381,6 +389,15 @@ def shots(self) -> Shots:
         """
         return self._shots
 
+    @property
+    def num_preps(self) -> int:
+        """Returns the index of the first operator that is not an StatePrepBase operator."""
+        idx = 0
+        num_ops = len(self.operations)
+        while idx < num_ops and isinstance(self.operations[idx], qml.operation.StatePrepBase):
+            idx += 1
+        return idx
+
     ##### Update METHODS ###############
 
     def _update(self):
```

Comment:
Do we need to update the `num_preps` property here?

Reply:
The property is dynamic instead of cached.